core/src/main/java/tc/oc/pgm/stats/StatType.java

```diff
@@ -1,43 +1,83 @@
 package tc.oc.pgm.stats;
 
 import static net.kyori.adventure.text.Component.translatable;
-import static net.kyori.adventure.text.format.NamedTextColor.*;
+import static net.kyori.adventure.text.format.NamedTextColor.GREEN;
+import static net.kyori.adventure.text.format.NamedTextColor.RED;
+import static net.kyori.adventure.text.format.NamedTextColor.YELLOW;
 import static tc.oc.pgm.stats.StatsMatchModule.damageComponent;
 import static tc.oc.pgm.util.text.NumberComponent.number;
 
 import java.util.Locale;
 import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.TextColor;
+import org.bukkit.Color;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.util.math.Formula;
 
-public enum StatType {
-  KILLS,
-  DEATHS,
-  ASSISTS,
-  KILL_STREAK,
-  BEST_KILL_STREAK,
-  KILL_DEATH_RATIO,
-  LONGEST_BOW_SHOT {
-    private final String blocks = key + ".blocks";
+public sealed interface StatType<I> {
+  Component makeNumber(Number number);
+
+  Component component(Component valueComponent);
+
+  Component component(I input);
+
+  enum Builtin implements StatType<StatHolder> {
+    KILLS,
+    DEATHS,
+    ASSISTS,
+    KILL_STREAK,
+    BEST_KILL_STREAK,
+    KILL_DEATH_RATIO,
+    LONGEST_BOW_SHOT {
+      private final String blocks = key + ".blocks";
+
+      @Override
+      public Component makeNumber(Number number) {
+        return translatable(blocks, number(number, YELLOW));
+      }
+    },
+    DAMAGE {
+      @Override
+      public Component makeNumber(Number number) {
+        return damageComponent(number.doubleValue(), GREEN);
+      }
+    };
+
+    public final String key = "match.stats.type." + name().toLowerCase(Locale.ROOT);
 
     @Override
     public Component makeNumber(Number number) {
-      return translatable(blocks, number(number, YELLOW));
+      return number(number, this == DEATHS ? RED : GREEN);
     }
-  },
-  DAMAGE {
+
     @Override
-    public Component makeNumber(Number number) {
-      return damageComponent(number.doubleValue(), GREEN);
+    public Component component(StatHolder stats) {
+      return translatable(key, makeNumber(stats.getStat(this)));
     }
-  };
-
-  public final String key = "match.stats.type." + name().toLowerCase(Locale.ROOT);
-  public static final StatType[] VALUES = values();
 
-  public Component makeNumber(Number number) {
-    return number(number, this == DEATHS ? RED : GREEN);
+    @Override
+    public Component component(Component valueComponent) {
+      return translatable(key, valueComponent);
+    }
   }
 
-  public Component component(StatHolder stats) {
-    return translatable(key, makeNumber(stats.getStat(this)));
+  record OfFormula(Component name, Formula<MatchPlayer> formula, Color color)
```

Comment:
```suggestion
  record OfFormula(Component name, Formula<MatchPlayer> formula, TextColor color)
```
The datatype here is most definitely not `Color`, those are arbitrary colors and are only used for leather armor.
RGB chat colors are modern and pgm hasn't adapted to them, the codebase relies in bukkit's `ChatColor` in many places still, and more recently, in `NamedTextColor`. Migrating those types eventually is a task we'll probs have to look at some day, but regardless, Color is not the right datatype even for generic color that supports either named or not, the type should be adventure's `TextColor`.

Reply:
Cool was going to make it a ChatColor but didn't see an existing way to convert it to a native TextColor, I see there's `TextFormatter#convert` as you've pointed out so can use that